core/primitives/src/reed_solomon.rs

```diff
@@ -54,14 +54,24 @@ pub fn reed_solomon_decode<T: BorshDeserialize>(
         return Err(Error::other(err));
     }
 
-    let encoded_data = parts
-        .iter()
-        .flat_map(|option| option.as_ref().expect("Missing shard").iter())
-        .cloned()
-        .take(encoded_length)
-        .collect_vec();
+    // Copy filled parts in chunks into a preallocated buffer,
+    // then truncate to the exact encoded length to remove padding.
+    let mut buf = Vec::with_capacity(encoded_length);
+    for part_opt in parts.iter() {
+        let part = part_opt.as_ref().expect("Missing shard");
+        if buf.len() >= encoded_length {
+            break;
+        }
```

Comment:
Is this condition right? Why is it necessary? Wouldn't the next snippet of code fill the buffer up-to `encoded_length` only anyway?

Reply:
It is not necessary for correctness

it was meant to be an early exit with no measurable speed up, I'll remove it 👍